File: Algo/Medium/1328_Break_a_Palindrome/Solution.py
## 2022-10-09 (unoptimized)
class Solution:
    def breakPalindrome(self, palindrome: str) -> str:
        letters = list(palindrome)
        
        ## search left to right for a char to change into 'a'
        for i in range(len(palindrome)//2):
            if palindrome[i] != 'a':
                palindrome[i] = 'a'
                return "".join(letters)

        ## search right to left for a char to increment
        for i in range(len(letters)-1, -1, -1):
            if letters[i] != 'z':
                letters[i] = chr(ord(letters[i])+1)
                return "".join(letters)

        return ""
        
## 2022-10-09 (optimized)
class Solution:
    def breakPalindrome(self, palindrome: str) -> str:
        if len(palindrome) < 2:
            return ""

        letters = list(palindrome)
        
        ## search left to right for a char to change into 'a'
        for i in range(len(letters)//2):
            if letters[i] != 'a':
                letters[i] = 'a'
                return "".join(letters)

        letters[len(palindrome)-1] = 'b'
        return "".join(letters)
        

File: Algo/Medium/1328_Break_a_Palindrome/Solution.cpp
// 2022-10-09
class Solution {
public:
    string breakPalindrome(string palindrome) {
        if (palindrome.length() < 2) return "";

        int n = palindrome.length();
        // only check half since if first half doesn't have an 'a', then
        // the second wouldn't either since it's a palindrome
        for (int i=0; i<n/2; i++) {
            if (palindrome[i] != 'a') {
                palindrome[i] = 'a';
                return palindrome;
            }
        }
        // if palindrome is entirely 'a's, then optimal solution would be to
        // change the last value into a 'b'
        palindrome[n-1] = 'b';
        return palindrome;
    }
};